### ISSUE TITLE
Intermittent AccessViolation Exceptions

### ISSUE THREAD

**Issue by thalter** [i-e-b/DiskQueue#26]
I'm running DiskQueue in a .NET Framework 4.8.0 application built as a Windows service. I'm collecting hundreds of thousands of events per day (collecting streaming data from a PLC), but it seems that once a day or so the application will terminate with a low-level Win32 exception. 

I'm running with the following settings:
ParanoidFlushing=true
AllowTruncatedEntres=true (I've also tried false and it didn't make any difference)
ThrowOnConflict=false (I've also tried true and it didn't make any difference)

I'm using separate threads (in the same process) for queuing and dequeuing. I have a single global PersistentQueue object, and each thread creates, flushes, and disposes of its own PersistentQueueSession object.

This isn't a super big deal, as I have the service configured to restart, but every time this happens (roughly once a day), I lose about 30 seconds of data capture.

Below is the stack trace from the Windows Application log.

```
Application: ScannerCollectorService.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException
   at Microsoft.Win32.Win32Native.MoveFile(System.String, System.String)
   at System.IO.File.InternalMove(System.String, System.String, Boolean)
   at System.IO.File.Move(System.String, System.String)
   at DiskQueue.Implementation.StandardFileDriver.Move(System.String, System.String)
   at DiskQueue.Implementation.StandardFileDriver.PrepareDelete(System.String)
   at DiskQueue.Implementation.StandardFileDriver.WaitDelete(System.String)
   at DiskQueue.Implementation.StandardFileDriver.AtomicWriteInternal(System.String, System.Action`1<System.IO.FileStream>)
   at DiskQueue.Implementation.StandardFileDriver.AtomicWrite(System.String, System.Action`1<DiskQueue.IBinaryWriter>)
   at DiskQueue.Implementation.PersistentQueueImpl.FlushTrimmedTransactionLog()
   at DiskQueue.Implementation.PersistentQueueImpl.CommitTransaction(System.Collections.Generic.ICollection`1<DiskQueue.Implementation.Operation>)
   at DiskQueue.Implementation.PersistentQueueSession.Flush()
   at ScannerCollectorService.ScannerDataCollector.queueReader_DoWork(System.Object, System.ComponentModel.DoWorkEventArgs)
   at System.ComponentModel.BackgroundWorker.OnDoWork(System.ComponentModel.DoWorkEventArgs)
   at System.ComponentModel.BackgroundWorker.WorkerThreadStart(System.Object)
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr, System.Object[], System.Object, System.Object[] ByRef)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.AsyncProcessMessage(System.Runtime.Remoting.Messaging.IMessage, System.Runtime.Remoting.Messaging.IMessageSink)
   at System.Runtime.Remoting.Proxies.AgileAsyncWorkerItem.DoAsyncCall()
   at System.Runtime.Remoting.Proxies.AgileAsyncWorkerItem.ThreadPoolCallBack(System.Object)
   at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by thalter** [comment]
A few more comments: System.AccessViolationException is a low-level Win32 exception that bypass the .NET exception handling, and hence cannot be caught and will cause the application to terminate. 

.NET Framework allows you to apply the [HandleProcessCorruptedStateExceptions](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.handleprocesscorruptedstateexceptionsattribute?view=netframework-4.8) attribute to your code and catch these exceptions, but it is not recommended by Microsoft, and I'm a little scared to try this since your application is now running in an inconsistent state.

**Comment by i-e-b** [comment]
Hi. Sorry for the late reply.
Can you tell me which version of DiskQueue you are using?

**Comment by thalter** [comment]
1.6.5. 

A few other observations:

My application has to be built as a 32-bit application due to the fact that my vendor library is 32-bit. There may be a resource leak somewhere in DiskQueue - my memory usage for my process regularly jumps up 4GB, the max for a 32-bit process, which could be the root cause for this crash. I switched to a .NET ConcurrentQueue, and while I may have lost disk-based resiliency, my memory usage has gone way down, and I haven't had any failures.

**Comment by i-e-b** [comment]
Could you please try using 1.6.7, and let me know if that helps?

Versions before 1.6.7 had a "feature" where they tried to give permissions for the `Everyone` user to access all queue files. This can cause various problems depending on the specific Windows environment.
This was turned off by default in 1.6.7.

There was also a bug with flushing and disposing of streams that was fixed in 1.6.6 that might be the cause of your memory use issue.

**Comment by i-e-b** [comment]
Going to close for now.
If you get the same issue with 1.6.7 or 1.6.8, please feel free to re-open.
